picoquic/sender.c

```diff
@@ -264,17 +264,32 @@ uint32_t picoquic_pad_to_policy(picoquic_cnx_t * cnx, uint8_t * bytes, uint32_t
  * Packet management
  */
 
-picoquic_packet_t* picoquic_create_packet()
+picoquic_packet_t* picoquic_create_packet(picoquic_quic_t * quic)
 {
-    picoquic_packet_t* packet = (picoquic_packet_t*)malloc(sizeof(picoquic_packet_t));
+    picoquic_packet_t* packet = quic->p_first_packet;
+    
+    if (packet == NULL) {
+        packet = (picoquic_packet_t*)malloc(sizeof(picoquic_packet_t));
+    }
+    else {
+        quic->p_first_packet = packet->next_packet;
+    }
 
     if (packet != NULL) {
-        memset(packet, 0, sizeof(picoquic_packet_t));
+        memset(packet, 0, offsetof(struct st_picoquic_packet_t, bytes));
     }
 
     return packet;
 }
 
+void picoquic_recycle_packet(picoquic_quic_t * quic, picoquic_packet_t* packet)
+{
+    if (packet != NULL) {
```

Comment:
Should we maybe introduce an upper bound for the number of packets? Otherwise we maybe create some sort of memory leak?

Reply:
Yes, but the question then is where to place the upper bound. Assume 1Gbps connections, 100ms RTT, you need close to 10000 packets. When the connection is up, these packets won't be in the pool, they will be in the retransmit queues. When a connection closes they will move to the pool, ready for the next one. But then how many connections are there?

I suppose we could have an arbitrary limit, like 16K packets or maybe 32K.

Reply:
OK, done.